### ISSUE TITLE
TypeError during the state_average calculation

### ISSUE THREAD

**Issue by jamcas** [pyscf/pyscf#335]
```
from pyscf import gto, scf, mcscf, qmmm, dmrgscf, molden, mrpt, lib
import numpy, os

dmrgscf.settings.MPIPREFIX = 'mpirun -np 10'

os.system('date')

mol = gto.M(atom="N 0 0 0; N 0 0 1", basis="ccpvdz")
mol.verbose = 4
mol.build()
mf = scf.RHF(mol)
mf.__dict__.update(lib.chkfile.load('test.chk', 'scf'))

mc = dmrgscf.DMRGSCF(mf, 4, 4)
mc.state_average_([0.5, 0.5])
mc.run()
mc.analyze()

Traceback (most recent call last):
  File "test.py", line 31, in <module>
    mc.analyze()
  File "/home/chenh/anaconda3/lib/python3.6/site-packages/pyscf/mcscf/casci.py", line 91, in analyze
    casdm1a, casdm1b = casscf.fcisolver.make_rdm1s(civec, ncas, nelecas)
  File "/home/chenh/anaconda3/lib/python3.6/site-packages/pyscf/mcscf/addons.py", line 650, in make_rdm1s
    dm1s = fcibase_class.make_rdm1s(self, ci0[i], norb, nelec, *args, **kwargs)
  File "/home/chenh/anaconda3/lib/python3.6/site-packages/pyscf/dmrgscf/dmrgci.py", line 287, in make_rdm1s
    dm1, dm2 = self.make_rdm12(state, norb, nelec, link_index, **kwargs)
  File "/home/chenh/anaconda3/lib/python3.6/site-packages/pyscf/mcscf/addons.py", line 658, in make_rdm12
    dm1, dm2 = fcibase_class.make_rdm12(self, ci0[i], norb, nelec, *args, **kwargs)
TypeError: 'int' object is not subscriptable
```
It works fine in CASSCF  but not in DMRGSCF.

**Comment by sanshar** [comment]
HI,
  Looks like it is calling casci rdms instead of dmrgci rdms, which is what is causing the failure. I will try to fix it today.

**Comment by jamcas** [comment]
@sanshar ,thanks for your help ! And i also found another error:
```
import os
from pyscf import gto, scf, mcscf, dmrgscf, lib, mrpt, qmmm
import sys
from pyscf.tools import molden
from pyscf import dft
import numpy

dmrgscf.settings.MPIPREFIX = 'mpirun -np 10'

os.system('date')

mol = gto.Mole()
mol.verbose = 4
mol.atom = open('S0-A-10o-10e-3.xyz').read()
mol.basis = 'ccpvdz'
mol.spin = 0
mol.charge = 0
mol.max_memory = 60000
mol.build()

# dft
mf = dft.RKS(mol)
mf.max_cycle = 500
mf.xc = 'b3lypg'
mf.__dict__.update(lib.chkfile.load('S0-A-10o-10e-3-dft.chk', 'scf'))

mc = mcscf.CASSCF(mf, 10, 10)
mc.conv_tol = 1e-5
mc.conv_tol_grad = 0.1
mc.chkfile = 'S0-A-10o-10e-3-state_average-spin=2.chk'
mc.state_average_([0.25, 0.25, 0.25, 0.25])
mc.fcisolver.spin = 2
mo = lib.chkfile.load('S0-A-10o-10e-3-cas.chk', 'mcscf/mo_coeff')
mc.run(mo)

mo_coeff_nature, ci_nature, occ_nature = mc.cas_natorb(sort=True)
mcscf.chkfile.dump_mcscf(mc, mo_coeff=mo_coeff_nature, mo_occ=occ_nature)
molden.from_mo(mol, 'S0-A-10o-10e-3-state_average-spin=2.molden', mo_coeff=mo_coeff_nature, ene=None, occ=occ_nature)

mc.analyze()

Traceback (most recent call last):
  File "S0-A-10o-10e-3.py", line 42, in <module>
    mo_coeff_nature, ci_nature, occ_nature = mc.cas_natorb(sort=True)
  File "/home/chenh/anaconda3/lib/python3.6/site-packages/pyscf/mcscf/casci.py", line 857, in cas_natorb
    with_meta_lowdin)
  File "/home/chenh/anaconda3/lib/python3.6/site-packages/pyscf/mcscf/casci.py", line 278, in cas_natorb
    for x in ci]
  File "/home/chenh/anaconda3/lib/python3.6/site-packages/pyscf/mcscf/casci.py", line 278, in <listcomp>
    for x in ci]
  File "/home/chenh/anaconda3/lib/python3.6/site-packages/pyscf/fci/addons.py", line 746, in transform_ci_for_orbital_rotation
    ci = lib.dot(trans_ci_a.T, ci.reshape(na,nb))
ValueError: cannot reshape array of size 44100 into shape (252,252)

```
### S0-A-10o-10e-3.xyz : 

> H                 -3.89002200    1.65877700    0.21940400
> H                 -2.09418700    6.17579300   -0.28769500
> H                 -4.54800100    5.89613200   -0.16358400
> H                 -5.48586400    3.58613300    0.09764000
> H                 -0.00999600    6.23188200   -0.36352600
> H                  2.46140800    6.13208900   -0.44391600
> H                  2.15284800    1.87091700   -0.04208900
> H                  3.57346100    3.88726400   -0.27743400
> H                  0.80020900    0.11338800    2.31000800
> H                  0.63010200   -0.23214200   -1.96613500
> H                  2.76237000   -1.35877800    2.34657400
> H                  2.59136700   -1.70556100   -1.92068100
> H                  3.68280800   -2.27603900    0.23302400
> C                 -3.54733900    2.68010100    0.10776400
> C                 -1.70549000    4.06481300   -0.09889200
> C                 -2.52178300    5.19053200   -0.17537300
> C                 -3.89713400    5.03331000   -0.10564100
> C                 -4.41978700    3.75648500    0.03887600
> C                 -0.23557100    4.10502000   -0.15948800
> C                  0.49664200    5.28076000   -0.29464500
> C                  1.87999700    5.22542000   -0.33970800
> C                  1.71223700    2.85320700   -0.11602400
> C                  2.49742200    3.98725900   -0.24825800
> C                  0.57990200    0.08744600    0.16541700
> C                  1.19359000   -0.26141200    1.37166100
> C                  1.09844500   -0.45465700   -1.01380300
> C                  2.30556800   -1.10157400    1.39747400
> C                  2.20989900   -1.29571200   -0.99214700
> C                  2.82197200   -1.61891800    0.21416600
> N                 -2.22210000    2.82827200    0.04038400
> N                  0.37328600    2.89747900   -0.07145500
> Ni                -0.82641100    1.35721200    0.11682400
> Br                -2.32304400   -0.40364100    0.34701400
Do you have any ideas about this ?

**Comment by sunqm** [comment]
@jamcas The problem is caused by spin. The statement mc.fcisolver.spin=2 changes the spin for fci solver but the spin for mc object is still 0. 

A workaround for now is to change the mc object to

mc = mcscf.CASSCF(mf, 10, (6,4))

I'll fix this problem in next release.

**Comment by jamcas** [comment]
@sunqm , This is really useful，but when I proceed to do the geometry optimization, another error has appeared:
```
import os
from pyscf import gto, scf, mcscf, dmrgscf, lib, mrpt, qmmm,geomopt
import sys
from pyscf.tools import molden
from pyscf import dft
import numpy

dmrgscf.settings.MPIPREFIX = 'mpirun -np 10'

os.system('date')

mol = gto.Mole()
mol.verbose = 4
mol.atom = open('S0-A-10o-10e-3.xyz').read()
mol.basis = 'ccpvdz'
mol.spin = 0
mol.charge = 0
mol.max_memory = 60000
mol.build()

# dft
mf = dft.RKS(mol)
mf.max_cycle = 500
mf.xc = 'b3lypg'
mf.__dict__.update(lib.chkfile.load('S0-A-10o-10e-3-dft.chk', 'scf'))

mc = mcscf.CASSCF(mf, 10, (6,4))
mc.conv_tol = 1e-5
mc.conv_tol_grad = 0.1
mc.chkfile = 'S0-A-10o-10e-3-state_average-spin=2.chk'
mc.state_specific_(state=3)
mc.fcisolver.spin = 2
mo = lib.chkfile.load('S0-A-10o-10e-3-cas.chk', 'mcscf/mo_coeff')
mc.run(mo)

mo_coeff_nature, ci_nature, occ_nature = mc.cas_natorb(sort=True)
mcscf.chkfile.dump_mcscf(mc, mo_coeff=mo_coeff_nature, mo_occ=occ_nature)
molden.from_mo(mol, 'S0-A-10o-10e-3-state_average-spin=2.molden', mo_coeff=mo_coeff_nature, ene=None, occ=occ_nature)
mc.analyze()

# geomopt
excited_grad = mc.nuc_grad_method().as_scanner()
mol1 = geomopt.optimize(excited_grad)

Traceback (most recent call last):
  File "/home/chenh/anaconda3/lib/python3.6/site-packages/pyscf/lib/chkfile.py", line 158, in load_mol
    with h5py.File(chkfile, 'r') as fh5:
  File "/home/chenh/anaconda3/lib/python3.6/site-packages/h5py/_hl/files.py", line 269, in __init__
    fid = make_fid(name, mode, userblock_size, fapl, swmr=swmr)
  File "/home/chenh/anaconda3/lib/python3.6/site-packages/h5py/_hl/files.py", line 99, in make_fid
    fid = h5f.open(name, flags, fapl=fapl)
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "h5py/h5f.pyx", line 78, in h5py.h5f.open
OSError: Unable to open file (file signature not found)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "S0-A-10o-10e-3.py", line 52, in <module>
    mol1 = geomopt.optimize(excited_grad)
  File "/home/chenh/anaconda3/lib/python3.6/site-packages/pyscf/geomopt/__init__.py", line 21, in optimize
    return berny_solver.optimize(method, *args, **kwargs)
  File "/home/chenh/anaconda3/lib/python3.6/site-packages/pyscf/geomopt/berny_solver.py", line 164, in optimize
    **kwargs)[1]
  File "/home/chenh/anaconda3/lib/python3.6/site-packages/pyscf/geomopt/berny_solver.py", line 131, in kernel
    energy, gradients = g_scanner(mol)
  File "/home/chenh/anaconda3/lib/python3.6/site-packages/pyscf/grad/casscf.py", line 156, in __call__
    e_tot = mc_scanner(mol)
  File "/home/chenh/anaconda3/lib/python3.6/site-packages/pyscf/mcscf/mc1step.py", line 515, in __call__
    mf_scanner(mol)
  File "/home/chenh/anaconda3/lib/python3.6/site-packages/pyscf/scf/hf.py", line 1146, in __call__
    dm0 = self.from_chk(self.chkfile)
  File "/home/chenh/anaconda3/lib/python3.6/site-packages/pyscf/scf/hf.py", line 1423, in from_chk
    return self.init_guess_by_chkfile(chkfile, project)
  File "/home/chenh/anaconda3/lib/python3.6/site-packages/pyscf/scf/hf.py", line 1421, in init_guess_by_chkfile
    return init_guess_by_chkfile(self.mol, chkfile, project=project)
  File "/home/chenh/anaconda3/lib/python3.6/site-packages/pyscf/scf/hf.py", line 495, in init_guess_by_chkfile
    dm = uhf.init_guess_by_chkfile(mol, chkfile_name, project)
  File "/home/chenh/anaconda3/lib/python3.6/site-packages/pyscf/scf/uhf.py", line 82, in init_guess_by_chkfile
    chk_mol, scf_rec = chkfile.load_scf(chkfile_name)
  File "/home/chenh/anaconda3/lib/python3.6/site-packages/pyscf/scf/chkfile.py", line 25, in load_scf
    return load_mol(chkfile), load(chkfile, 'scf')
  File "/home/chenh/anaconda3/lib/python3.6/site-packages/pyscf/lib/chkfile.py", line 163, in load_mol
    with h5py.File(chkfile, 'r') as fh5:
  File "/home/chenh/anaconda3/lib/python3.6/site-packages/h5py/_hl/files.py", line 269, in __init__
    fid = make_fid(name, mode, userblock_size, fapl, swmr=swmr)
  File "/home/chenh/anaconda3/lib/python3.6/site-packages/h5py/_hl/files.py", line 99, in make_fid
    fid = h5f.open(name, flags, fapl=fapl)
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "h5py/h5f.pyx", line 78, in h5py.h5f.open
OSError: Unable to open file (file signature not found)
```
I think it is because I used mf .__ dict __. Update (lib.chkfile.load ( 'S0-A-10o-10e-3-dft.chk', 'scf')) resulting in chkfiles can not be read correctly. 
So I can only re-calculated once dft calculation instead of using mf .__ dict __. Update (lib.chkfile.load ( 'S0-A-10o-10e-3-dft.chk', 'scf')) to keep the optimization going on.

**Comment by sunqm** [comment]
Problems are solved in PR #339. It will be available in pyscf-1.7 release

**Comment by jamcas** [comment]
@sunqm, thank you very much for your help!
